tools/wptserve/wptserve/response.py

```diff
@@ -234,27 +235,46 @@ def write(self):
         self.write_status_headers()
         self.write_content()
 
-    def set_error(self, code, message=""):
+    def set_error(self, code, err=None):
         """Set the response status headers and return a JSON error object:
 
         {"error": {"code": code, "message": message}}
         code is an int (HTTP status code), and message is a text string.
         """
-        err = {"code": code,
-               "message": message}
-        data = json.dumps({"error": err})
+        if 500 <= code < 600:
+            message = self._format_server_error(err)
+            self.logger.warning(message)
+        else:
+            if err is None:
+                message = ""
+            else:
+                message = str(err)
+
+        data = json.dumps({"error": {
+            "code": code,
+            "message": message}
+        })
         self.status = code
         self.headers = [("Content-Type", "application/json"),
                         ("Content-Length", len(data))]
         self.content = data
-        if code == 500:
-            if isinstance(message, str) and message:
-                first_line = message.splitlines()[0]
-            else:
-                first_line = "<no message given>"
-            self.logger.error("Exception loading %s: %s" % (self.request.url,
-                                                            first_line))
-            self.logger.info(message)
+
+    def _format_server_error(self, err):
+        if err is None:
+            suffix = "<no traceback>"
+        elif isinstance(err, str):
+            suffix = err
+        elif self.request.server.config.logging["suppress_handler_traceback"]:
+            frame = traceback.extract_tb(err.__traceback__)[-1]
```

Comment:
Do we need the indirection of `traceback.extract_tb` here? Or am I just failing to understand all of the various traceback objects?

Reply:
Not sure, what do you think would work instead?

`extract_tb` gives us a `StackSummary` object which we can then use to get access to the last stack frame directly. It's possible we could also use one of the other methods to directly get a list of strings and then extract from that, but this at least gives us complete control.

Reply:
Yeah, maybe nothing else would work, digging into the various APIs. I thought we could get this directly from `__traceback__`, but it seems like probably not.